okhttp/src/main/java/okhttp3/CertificatePinner.kt

```diff
@@ -239,22 +241,22 @@ class CertificatePinner internal constructor(
   }
 
   internal data class Pin(
-    /** A hostname like `example.com` or a pattern like `*.example.com`. */
-    val pattern: String,
-    /** The canonical hostname, i.e. `EXAMPLE.com` becomes `example.com`. */
-    private val canonicalHostname: String,
+    /** A hostname like `example.com` or a pattern like `*.example.com` (canonical form). */
+    private val pattern: String,
     /** Either `sha1/` or `sha256/`. */
     val hashAlgorithm: String,
     /** The hash of the pinned certificate using [hashAlgorithm]. */
     val hash: ByteString
   ) {
+
     fun matches(hostname: String): Boolean {
-      if (pattern.startsWith(WILDCARD)) {
-        val firstDot = hostname.indexOf('.')
-        return hostname.length - firstDot - 1 == canonicalHostname.length &&
-            hostname.startsWith(canonicalHostname, startIndex = firstDot + 1)
+      val ending by lazy { pattern.substringAfter(".") }
```

Comment:
Having lazy inside the method seems like non idiomatic kotlin to me. I'd suggest another round of cleanup of the logic. I think it's right, and I'm not against landing, but I suspect there is a nicer succinct version of this hidden here.

One suggestion, TIOLI: maybe pulling out this wildcard matching into a internal sealed class that you test separately from Pin?